bgpv2: container labs for various types of advertisements

harsimran-pabla opened the pull request:
Adding container labs for local development and testing of BGPv2 APIs. 

PR is split by BGP advertisement type
- PodCIDR
- Multi-homing
- CiliumPodIPPool
- Service types : Loadbalancer, cluster IP, external IP

harsimran-pabla committed:
bgpv2: container lab example for multi-homing bgp

Adding lab to demonstrate two BGP neighbors from same node to which pod
CIDR prefixes are advertised.

Signed-off-by: harsimran pabla <hpabla@isovalent.com>

harsimran-pabla committed:
bgpv2: container lab example for CiliumPodIPPool

Adding lab to demonstrate advertisement of CiliumPodIPPool using BGPv2
APIs.

Signed-off-by: harsimran pabla <hpabla@isovalent.com>

harsimran-pabla committed:
bgpv2: container lab example for service types and pod cidr

Adding labs to show service advertisements for load-balancer, external
IP and cluster IP using BGPv2 APIs. Also adding example of PodCIDR
advertisement, each using different NLRI attributes.

Signed-off-by: harsimran pabla <hpabla@isovalent.com>

harsimran-pabla committed:
CODEOWNERS: assign contrib/containerlab to sig-bgp

Container lab directory contains only BGP specific labs. Assigning
directory code ownership to sig-bgp for time being.

Signed-off-by: harsimran pabla <hpabla@isovalent.com>

YutaroHayakawa commented on contrib/containerlab/bgpv2/pod-cidr/Makefile:
```diff
@@ -0,0 +1,20 @@
+	# create secret for bgp
+	kubectl -n kube-system create secret generic --type=string bgp-auth-secret --from-literal=password=cilium123
+	# install cilium
+	helm install cilium -n kube-system oci://quay.io/cilium-charts-dev/cilium --version 1.16.0-dev-dev.247-main-8c97a2119d -f values.yaml
```
I think we should be able to override this Cilium version.

harsimran-pabla replied:
Added an option to override cilium version. This is temporary, till official release is done. At that point, I would just remove specific cilium version installation and get latest.

YutaroHayakawa reviewed (changes requested):
Nice work! I think we can merge PodCIDR and Service labs into the single lab as their ContainerLab and Cilium configurations are the same.

bimmlerd commented on contrib/containerlab/bgpv2/multi-homing/bgp.yaml:
```diff
@@ -0,0 +1,59 @@
+    - advertisementType: "PodCIDR"
+      attributes:
+        communities:
+          wellKnown: [ "no-export" ]
```
nit: no newline at EOF

bimmlerd commented on contrib/containerlab/bgpv2/multi-homing/Makefile:
```diff
@@ -0,0 +1,20 @@
+	# create secret for bgp
+	kubectl -n kube-system create secret generic --type=string bgp-auth-secret --from-literal=password=cilium123
+	# install cilium
+	helm install cilium -n kube-system oci://quay.io/cilium-charts-dev/cilium --version 1.16.0-dev-dev.247-main-8c97a2119d -f values.yaml
```
similarly here, this is a very specific cilium version

harsimran-pabla replied:
Made this configurable, waiting for next official release at which point I will remove it and get latest.

YutaroHayakawa replied:
How did you make it configurable? I don't see the knob. Am I missing something?

harsimran-pabla replied:
woops, commit got lost somewhere  :( checking

harsimran-pabla replied:
Pushed the change again to have it fixed.

bimmlerd commented on contrib/containerlab/bgpv2/multi-homing/Makefile:
```diff
@@ -0,0 +1,20 @@
+deploy:
+	kind create cluster --config cluster.yaml
+	sudo containerlab -t topo.yaml deploy
```
could add a check for containerlab to print a more helpful message from where to install it if one doesn't yet have it

harsimran-pabla replied:
I wanted to keep these make files simple. I added a readme section under containerlab/bgpv2, which provides prerequisites of installing kind, containterlab etc. Hope that is enough for users to know what kind of binaries are required to run the lab. I think target audience probably already have container lab - if not, readme should help.

bimmlerd reviewed (changes requested):
I'm a bit lost as to what these are - could we add a bit of documentation somewhere?

harsimran-pabla committed:
bgpv2: adding readme files for each lab

Adding readme files to provide pre-requisites required to run container
lab kind kubernetes cluster. As well as documentation about individual
lab.

Signed-off-by: harsimran pabla <hpabla@isovalent.com>
